cs169/BJC-Teacher-Tracker-App
# 187384622/new email model - PR1 - Change Email and Make Sure original tests didn't break

### [Pivotal Tracker Link1][tracker1]


[tracker1]: https://www.pivotaltracker.com/story/show/187384622

## What This PR Does

This pull request is the first in a series intended to facilitate easier code review. It restructures the `teacher` model to enable a `teacher` to have multiple `email_addresses`. The primary goals of this feature are:
- To allow administrators to manually link several email addresses to a teacher, while a teacher can only modify their primary email (typically a school email).
- To split the implementation across three PRs for clarity and manageability. This initial PR focuses on backend model changes, updates to value access in the frontend template and controller, modifications to existing tests, and adjustments to some seed data and fixtures.

Notably, this PR does not introduce new tests. Its primary objective is to ensure that existing tests continue to pass without issues. Subsequent PRs will include:
- A library task script for migrating current data (not schema) to align with the updated data model, accompanied by an RSpec test.
- Frontend enhancements allowing administrators to add multiple personal emails for a teacher and displaying these emails for both administrators and teachers, complete with comprehensive testing.

### Screenshots and Visuals

After running `rails db:migrate` and `rails db:seed`, the following changes are observable:
1. Teacher names and other associated attributes remain visible, except for the `email` field for original entries in the database. This is due to the migration of the database schema without corresponding data migration, leaving the new email model without the original email information. A forthcoming PR will address this with a data migration script.
2. The current frontend does not yet support displaying multiple email addresses and instead shows a primary email alongside an array of non-primary emails for each teacher.
<img width="1547" alt="image" src="https://github.com/cs169/BJC-Teacher-Tracker-App/assets/92573736/4555b171-e1df-4a76-8f2c-4bb057e6b163">
<img width="1547" alt="image" src="https://github.com/cs169/BJC-Teacher-Tracker-App/assets/92573736/a639b061-d733-4dcf-bc22-3fcf40170010">


#### Who authored this PR?

1. Perry (Jingchao) Zhong (@perryzjc)
4. Michael Tao (@realmichaeltao )


### How should this PR be tested?
At least make sure every original feature doesn't break.

* Is there a deploy we can view?
* What do the specs/features test?
Specs & Features are basically same as the original tests, with changes only relevant to some keywords changing & the way mocking the data (such as originally we can directly create a teacher with a email, now we need to associate them together)
* Are there edge cases to watch out for?
For this PR, at least make sure it  didn't break any existing test.

#### Are there any complications to deploying this?


1. `rails db:migrate`
2. `rails db:seed`

### Checklist:
